### PR TITLE
Proxy refactoring

### PR DESCRIPTION
**Pull request by bex1**
All proxies refactored.

The routes are grouped by the type of the respond result.

Adds compile time errors for invalid posts and updates etc. Can be hooked up to generated interfaces from the API later.

All routes in the API is covered. Except like 2-4 non relevant ones.

The proxy methods are consistent throughout which was not the case before.

### CODE REVIEW

**Comment by bex1** [comment]
Also switched to system_language_id for users @buren

**Comment by bex1** [comment]
@buren Remind me to go through some small questions I have regarding this PR with you on Wednesday. Its nothing major.

File: src/app/models/api-models/application/application.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ interface ApplicationApiAttributes {
   invoice: Invoice;
   job: Job;
   jobEnded: boolean;
-  language: Language;
+  language: Language; // TODO missing relation in docs?
```
Comment (buren) [inline src/app/models/api-models/application/application.ts]: What do you mean missing relation? :) I see it here: https://api.justarrived.se/api_docs/1.0/job_users.html

File: src/app/models/api-models/chat/chat.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import {UserImage} from '../user-image/user-image';
+import {UserImageFactory} from '../user-image/user-image';
+
+// API attribute interfaces
+interface ChatApiAttributes {
+  createdAt: Date;
+  id: string;
+  messages: Message[];
+  userImages: UserImage[];
+  users: User[];
```
Comment (buren) [inline src/app/models/api-models/chat/chat.ts:15]: 🙇 ❤️

File: src/app/models/api-models/country/country.ts [hunk 3]
```diff
@@ -6,7 +6,7 @@ interface CountryApiAttributes {
   countryCode: string;
   id: string;
   languageId: string;
-  language: Language;
+  language: Language; // TODO missing relation in docs?
```
Comment (buren) [inline src/app/models/api-models/country/country.ts]: ?

File: src/app/models/api-models/faq/faq.ts [hunk 4]
```diff
@@ -6,7 +6,7 @@ interface FaqApiAttributes {
   answer: string;
   category: string;
   id: string;
-  language: Language;
+  language: Language; // TODO missing relation in docs?
```
Comment (buren) [inline src/app/models/api-models/faq/faq.ts]: ?

File: src/app/models/api-models/job-language/job-language.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import {Job} from '../job/job';
+import {JobFactory} from '../job/job';
+import {Language} from '../language/language';
+import {LanguageFactory} from '../language/language';
+
+// TODO verify that the spec looks like below
```
Comment (buren) [inline src/app/models/api-models/job-language/job-language.ts]: You're right I does look like this! 😄

File: src/app/models/api-models/language/language.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@
 interface LanguageApiAttributes {
   direction: string;
   id: string;
+  language: Language; // TODO missing relation in docs?
```
Comment (buren) [inline src/app/models/api-models/language/language.ts]: ?

File: src/app/models/api-models/user-gender/user-gender.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import {Language} from '../language/language';
+import {LanguageFactory} from '../language/language';
+
+// API attribute interfaces
+interface UserGenderApiAttributes {
+  id: string;
+  name: string;
+  language: Language; // TODO missing relation in docs?
+  languageId: string; // TODO missing in docs?
```
Comment (buren) [inline src/app/models/api-models/user-gender/user-gender.ts]: Ah, this is a good opportunity to discuss this :) 

These resources are persisted in the database and these strings are loaded from a YAML-file. So they're kinda guaranteed to be translated :) however I can really see a quite strong case for having translations be uniform across all of our API-resources.

File: src/app/models/api-models/user-image-category/user-image-category.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import {Language} from '../language/language';
+import {LanguageFactory} from '../language/language';
+
+// API attribute interfaces
+interface UserImageCategoryApiAttributes {
+  description: string;
+  id: string;
+  language: Language; // TODO missing relation in docs?
```
Comment (buren) [inline src/app/models/api-models/user-image-category/user-image-category.ts]: Same as the comment above.

File: src/app/models/api-models/user-status/user-status.ts [hunk 9]
```diff
@@ -2,31 +2,31 @@ import {Language} from '../language/language';
 import {LanguageFactory} from '../language/language';
 
 // API attribute interfaces
-interface StatusApiAttributes {
+interface UserStatusApiAttributes {
   description: string;
   id: string;
-  language: Language;
+  language: Language; // TODO missing relation in docs?
```
Comment (buren) [inline src/app/models/api-models/user-status/user-status.ts]: See earlier comment

File: src/app/proxies/user/user.proxy.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+  consent: boolean;
+  country_of_origin?: string;
+  current_status?: string;
+  description?: string;
+  education?: string;
+  email: string;
+  first_name: string;
+  gender?: string;
+  ignored_notifications?: string[];
+  interest_ids?: {id: string, proficiency: number}[]; // TODO check why required in docs.
```
Comment (buren) [inline src/app/proxies/user/user.proxy.ts]: ❓
screen shot 2017-03-28 at 09 25 01

File: src/app/proxies/user/user.proxy.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  job_experience?: string;
+  language_id?: string;
+  language_ids?: {id: string, proficiency: number}[]; // TODO check why required in docs.
+  last_name: string;
+  next_of_kin_name?: string;
+  next_of_kin_phone?: string;
+  password?: string;
+  phone?: string;
+  skill_ids?: {id: string, proficiency: number}[]; // TODO check why required in docs.
```
Comment (buren) [inline src/app/proxies/user/user.proxy.ts]: ❓
screen shot 2017-03-28 at 09 25 13

File: src/app/proxies/user/user.proxy.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+  description?: string;
+  education?: string;
+  email: string;
+  first_name: string;
+  gender?: string;
+  ignored_notifications?: string[];
+  interest_ids?: {id: string, proficiency: number}[]; // TODO check why required in docs.
+  job_experience?: string;
+  language_id?: string;
+  language_ids?: {id: string, proficiency: number}[]; // TODO check why required in docs.
```
Comment (buren) [inline src/app/proxies/user/user.proxy.ts]: ❓
screen shot 2017-03-28 at 09 26 21

**Review by buren — COMMENTED** [review]
Wow! Just wow!
🙇

Amazing that you've managed to walkthrough, not just some of it, but all of it! ⚡️

File: src/app/proxies/user/user.proxy.ts (hunk 12 above)
Comment (bex1) [reply src/app/proxies/user/user.proxy.ts]: @buren 
image

Comment (bex1) [reply src/app/proxies/user/user.proxy.ts]: https://api.justarrived.se/api_docs/1.0/users/create.html

File: src/app/proxies/user/user.proxy.ts (hunk 10 above)
Comment (bex1) [reply src/app/proxies/user/user.proxy.ts]: You are right here 👍 Didn't look close enough

File: src/app/proxies/user/user.proxy.ts (hunk 11 above)
Comment (bex1) [reply src/app/proxies/user/user.proxy.ts]: You are right here 👍 Didn't look close enough

File: src/app/models/api-models/user-gender/user-gender.ts (hunk 7 above)
Comment (bex1) [reply src/app/models/api-models/user-gender/user-gender.ts]: Yeah thats what i thought aswell. All with translated_text could have a language_id and a language relation to make it consistent @buren 😄

File: src/app/models/api-models/job-language/job-language.ts (hunk 5 above)
Comment (bex1) [reply src/app/models/api-models/job-language/job-language.ts]: Cheers @buren

File: src/app/models/api-models/application/application.ts (hunk 1 above)
Comment (bex1) [reply src/app/models/api-models/application/application.ts]: Not under relationships in the route docs. 😄 @buren The language_id is there though

**Pull request merged**